### PR TITLE
Honor IDE Proxy configuration. (#149)

### PR DESCRIPTION
**Pull request by doublefx**
[kind/bug]

## What versions of software are you using?
- Operating System: Windows 7
- IntelliJ version: 2019.3
- OpenShift version: 3.11

## What did you run exactly?

- Set IntelliJ http proxy
- Install intellij-openshift-connector 

## Actual behavior
intellij-openshift-connector is proposing to download odo but fails silently.

## Expected behavior
intellij-openshift-connector honors IntelliJ http proxy and download odo

## Any logs, error output, etc?
nope

### CODE REVIEW

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 1]
```diff
@@ -190,7 +193,31 @@ private String getOdoCommand() throws IOException {
           command = ProgressManager.getInstance().run(new Task.WithResult<String, IOException>(null, "Downloading Odo", true) {
             @Override
             public String compute(@NotNull ProgressIndicator progressIndicator) throws IOException {
-              OkHttpClient client = new OkHttpClient();
+              OkHttpClient client;
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java]: Should be refactored as a separate method.

Comment (doublefx) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java]: Good point, will do tomorrow from home with my own github user.
Also, this class deserves a nice cleanup IMO, do you mind if I do it in another commit?

Comment (jeffmaury) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java]: I prefer to keep different concerns in different PRs so I would prefer another PR.
I you use a different GitHub id, don't forget the DCO (sign) header as this is required for us to accept contributions

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 2]
```diff
@@ -67,23 +68,22 @@
 import me.snowdrop.servicecatalog.api.model.ServiceInstance;
 import me.snowdrop.servicecatalog.api.model.ServiceInstanceFluent;
 import me.snowdrop.servicecatalog.api.model.ServiceInstanceList;
-import okhttp3.OkHttpClient;
-import okhttp3.Request;
-import okhttp3.Response;
+import okhttp3.*;
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java]: Starred import are forbidden by our policy

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java [hunk 3]
```diff
@@ -104,6 +104,8 @@
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
+import static com.intellij.openapi.util.text.StringUtil.isNotEmpty;
+import static java.net.Proxy.Type.*;
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java]: Idem

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java (hunk 2 above)
Comment (doublefx) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java]: Done!

File: src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java (hunk 3 above)
Comment (doublefx) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/odo/OdoCli.java]: Done!

**Comment by doublefx** [comment]
Also, I noticed the odo commands are executed by the DefaultExecutor.
Odo will not honor the IDE proxy but eventually the HTTP_PROXY, HTTPS_PROXY, NO_PROXY, ALL_PROXY environment variables if set.
Because of the way the IDE defines the proxy, there is no way obvious to determine whether it is an http or https proxy from it, therefore no obvious way to pass them to odo as environment variables using the DefaultExecutor.execute(final CommandLine command, final Map<String, String> environment) .

File: src/main/java/org/jboss/tools/intellij/openshift/actions/cluster/LoginAction.java [hunk 4]
```diff
@@ -33,9 +36,10 @@ public void actionPerformed(AnActionEvent anActionEvent, TreePath path, Object s
             return dialog;
             });
           if (loginDialog.isOK()) {
+            ExecHelper.setProxyEnvironmentVariables(OkHttpClientProxy.buildEnvironmentVariables(loginDialog.getClusterURL()));
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/actions/cluster/LoginAction.java]: No that should not be managed externally like that. The tree model has an refresh method that is called when the context is changed, that should be the starting point.

File: src/main/java/org/jboss/tools/intellij/openshift/utils/ExecHelper.java [hunk 5]
```diff
@@ -47,6 +48,12 @@
 public class ExecHelper {
     private static final ScheduledExecutorService SERVICE = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors());
 
+    private static Map<String, String> proxyEnvironmentVariables;
```
Comment (jeffmaury) [inline src/main/java/org/jboss/tools/intellij/openshift/utils/ExecHelper.java]: Don't like to much that ExecHelper has a state like that.

Comment (doublefx) [reply src/main/java/org/jboss/tools/intellij/openshift/utils/ExecHelper.java]: Me neither but I can't see a simpler option, a more complex would be to pass down the proxyEnvironmentVariables to the execute methods which implies a lot of refactoring and did not want to take that path.

Do you see another way?

File: src/main/java/org/jboss/tools/intellij/openshift/actions/cluster/LoginAction.java (hunk 4 above)
Comment (doublefx) [reply src/main/java/org/jboss/tools/intellij/openshift/actions/cluster/LoginAction.java]: 👍

**Pull request closed without merging**